src/RoboSaverVirtualModuleFactory.sol

```diff
@@ -0,0 +1,124 @@
+
+contract RoboSaverVirtualModuleFactory {
+    /*//////////////////////////////////////////////////////////////////////////
+                                     DATA TYPES
+    //////////////////////////////////////////////////////////////////////////*/
+    struct VirtualModuleDetails {
+        address virtualModuleAddress;
+        uint256 upkeepId;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                   CONSTANTS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    IERC20 constant LINK = IERC20(0xE2e73A1c69ecF83F464EFCE6A5be353a37cA09b2);
+    IKeeperRegistryMaster constant CL_REGISTRY = IKeeperRegistryMaster(0x299c92a219F61a82E91d2062A262f7157F155AC1);
+    IKeeperRegistrar constant CL_REGISTRAR = IKeeperRegistrar(0x0F7E163446AAb41DB5375AbdeE2c3eCC56D9aA32);
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                   PUBLIC STORAGE
+    //////////////////////////////////////////////////////////////////////////*/
+
+    // card -> (module address, upkeep id)
+    mapping(address => VirtualModuleDetails) public virtualModules;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                       EVENTS
+    //////////////////////////////////////////////////////////////////////////*/
+    event RoboSaverVirtualModuleCreated(address virtualModule, address card, uint256 upkeepId, uint256 timestamp);
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                       ERRORS
+    //////////////////////////////////////////////////////////////////////////*/
+    error UpkeepZero();
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                     CONSTRUCTOR
+    //////////////////////////////////////////////////////////////////////////*/
+
+    constructor() {
+        LINK.approve(address(CL_REGISTRAR), type(uint256).max);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                  EXTERNAL METHODS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @notice Creates a virtual module for a card
+    /// @param _delayModule The address of the delay module
+    /// @param _roleModule The address of the role module
+    /// @param _buffer The buffer for the virtual module (configurable)
+    /// @param _slippage The slippage for the virtual module (configurable)
+    function createVirtualModule(address _delayModule, address _roleModule, uint256 _buffer, uint16 _slippage)
+        external
+    {
+        // @todo sanity checks on the inputs!
+
+        // uses `CARD` address has to helps pre-determining the address of the virtual module given the salt
```

Comment:
this goes in line with the thought we discuss on how to allow creating the module and queue enabling atomically in a multicall at some point via the UI. w/o knowing the deterministic address of the module when it will get deploy it is imposible to queue at the same time `enableModule(<DETERMINISTIC_ADDY>)`. this approach try to tackle that